subsys/bluetooth/host/classic/a2dp.c

```diff
@@ -216,235 +178,212 @@ static int a2dp_set_config_ind(struct bt_avdtp *session, struct bt_avdtp_sep *se
-	}
+	__ASSERT(sep, "Invalid sep");
+	req_cb = a2dp_cb != NULL ? a2dp_cb->establish_req : NULL;
+	done_cb = (ep->stream != NULL && ep->stream->ops != NULL) ? ep->stream->ops->established
+								  : NULL;
+	return a2dp_ctrl_ind(session, sep, errcode, req_cb, done_cb, false);
+}
 
-	ops = stream->ops;
-	if (*errcode == 0) {
-		if ((ops != NULL) && (stream->ops->released != NULL)) {
-			stream->ops->released(stream);
-		}
-	}
+static int a2dp_start_ind(struct bt_avdtp *session, struct bt_avdtp_sep *sep, uint8_t *errcode)
+{
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+	bt_a2dp_ctrl_req_cb req_cb;
+	bt_a2dp_ctrl_done_cb done_cb;
 
-	return (*errcode == 0) ? 0 : -1;
+	__ASSERT(sep, "Invalid sep");
+	req_cb = a2dp_cb != NULL ? a2dp_cb->start_req : NULL;
+	done_cb = (ep->stream != NULL && ep->stream->ops != NULL) ? ep->stream->ops->started : NULL;
+	return a2dp_ctrl_ind(session, sep, errcode, req_cb, done_cb, false);
 }
 
 static int a2dp_suspend_ind(struct bt_avdtp *session, struct bt_avdtp_sep *sep, uint8_t *errcode)
 {
-	struct bt_a2dp_ep *ep;
-	struct bt_a2dp_stream *stream;
-	struct bt_a2dp_stream_ops *ops;
-
-	*errcode = 0;
-	ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
-	if (ep->stream == NULL) {
-		*errcode = BT_AVDTP_BAD_ACP_SEID;
-		return -1;
-	}
-	stream = ep->stream;
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+	bt_a2dp_ctrl_req_cb req_cb;
+	bt_a2dp_ctrl_done_cb done_cb;
 
-	if ((a2dp_cb != NULL) && (a2dp_cb->suspend_req != NULL)) {
-		uint8_t rsp_err_code;
-		int err;
-
-		err = a2dp_cb->suspend_req(stream, &rsp_err_code);
-		if (err) {
-			*errcode = rsp_err_code;
-		}
-	}
-
-	ops = stream->ops;
-	if (*errcode == 0) {
-		if ((ops != NULL) && (stream->ops->suspended != NULL)) {
-			stream->ops->suspended(stream);
-		}
-	}
-
-	return (*errcode == 0) ? 0 : -1;
+	__ASSERT(sep, "Invalid sep");
+	req_cb = a2dp_cb != NULL ? a2dp_cb->suspend_req : NULL;
+	done_cb =
+		(ep->stream != NULL && ep->stream->ops != NULL) ? ep->stream->ops->suspended : NULL;
+	return a2dp_ctrl_ind(session, sep, errcode, req_cb, done_cb, false);
 }
 
-static int bt_a2dp_open_cb(struct bt_avdtp_req *req)
+static int a2dp_close_ind(struct bt_avdtp *session, struct bt_avdtp_sep *sep, uint8_t *errcode)
 {
-	struct bt_a2dp *a2dp = OPEN_PARAM(OPEN_REQ(req));
-	struct bt_a2dp_ep *ep;
-	struct bt_a2dp_stream *stream;
-	struct bt_a2dp_stream_ops *ops;
+	struct bt_a2dp_ep *ep = CONTAINER_OF(sep, struct bt_a2dp_ep, sep);
+	bt_a2dp_ctrl_req_cb req_cb;
+	bt_a2dp_ctrl_done_cb done_cb;
 
-	ep = CONTAINER_OF(a2dp->open_param.sep, struct bt_a2dp_ep, sep);
-	if (ep->stream == NULL) {
-		return -1;
-	}
-	stream = ep->stream;
-
-	LOG_DBG("OPEN result:%d", a2dp->open_param.status);
+	__ASSERT(sep, "Invalid sep");
+	req_cb = a2dp_cb != NULL ? a2dp_cb->release_req : NULL;
+	done_cb =
+		(ep->stream != NULL && ep->stream->ops != NULL) ? ep->stream->ops->released : NULL;
+	return a2dp_ctrl_ind(session, sep, errcode, req_cb, done_cb, false);
```

Comment:
Whether the last parameter `clear_stream` should be `true`?

Reply:
Updated.